### ISSUE TITLE
Conditionals for different frameworks

### ISSUE THREAD

**Issue by astraldev** [BuilderIO/mitosis#1096]
### I am interested in helping provide a fix!

No

### Which generators are impacted?

- [X] All
- [ ] Angular
- [ ] HTML
- [ ] Qwik
- [ ] React
- [ ] React-Native
- [ ] Solid
- [ ] Stencil
- [ ] Svelte
- [ ] Vue
- [ ] Web components

### Reproduction case

https://mitosis.builder.io/?outputTab=I4dwlg1kA%3D%3D%3D&code=JYWwDg9gTgLgBAbzgVwM4FMDKMCGN0A0cARssADYAm6UcAvnAGZQQhwBEAAqRdVAHTAIAehDAYEVMFTsA3ACh56AB6RYcaoxzJy8RsgB2AYxhCDcALIBPAMKtIB9AZgAKMCzCoAlInlw4RhAGqPAA2gY4IIRwGDAAcpHoALpwALwoGNh46C7s2OgAbujsXgp%2BcFDoMMhQ5i7l%2FgA8lMAFAHwN%2FnCNwAZgyDCdXQGoqKkICEPDARDk0ABcHJWU7ART%2FnR063AFOOTI6OMRUVvTXUg8VDT8AOZVLgDkzIkA7tAA1g8%2BqT9L6DgmdhwAD8iAYiwQDFOZzgQRsAAscAY7uMXIUnDBvm0YlUElE0UVnPxcFA7jB%2BLt9ugvNDhsIOmcABLochzACEcAAkgEkRVDHBenAAEr%2FExEABqByImFmwEoRGgcAAMsAAI5kShszqNYQtdrlUryOhAA%3D%3D%3D

### Expected Behaviour

It would be nice if it were possible to check the framework a particular `.lite.tsx` file is being compiled into. 

### Actual Behaviour

Currently there is no way to check for that value.

### Additional Information

I would be nice to be able to use this 
```tsx
        ...
        { builder.get('framework') === react ? ...{ onClick: () => {} } : ...{ thisEventHandler: () => {} }
        ...
```

**Comment by Lokprakash-babu** [comment]
I would like to work on it @samijaber

**Comment by Lokprakash-babu** [comment]
From this line https://github.com/BuilderIO/mitosis/blob/6af5bd00c9056ee019a30d7062ffcb2c34cc3435/packages/cli/src/build/build.ts#L227, I can understand that Mitosis/Svelteosis JSX is parsed only once and cloned for each of the targets which generators can modify. We need to have a way for identifying the target while parsing the mitosis JSX. Also adding the framework getter as a part of builder parser a good approach, Since builder parser is for builder components I think..?

**Comment by samijaber** [comment]
Yeah so, this is definitely a feature we've wanted to have in Mitosis! The only question is how to go about it.

PS: This is not a simple feature! I'm doing my best to outline all the details, but I might be missing some ideas. Anyone interested is free to take a stab at it and ask for help along the way 🙏

What I currently envision is a `useTarget` hook that compiles away into a new `conditional` value in the JSON. Then, each generator will go through the JSON and look over the `conditional` blocks, and either preserve their content or strip them away if they are for a different `target`

## 1- add `useTarget` hook:

Add a `useTarget` hook that the user would import into their component:

```tsx
import { useState, useTarget, onUpdate } from '@builder.io/mitosis';

export default function MyComponent(props) {
  const [name, setName] = useState(
    useTarget({
      qwik: () => {
        return 'qwik' + '10';
      },
      react: () => 'react',
      default: () => 'other',
    })
  );


  useTarget({
    qwik: () => {
      onUpdate(() => {
        console.log('foo was updated')
      }, [props.foo])
    },
    default: () => undefined,
  })

  return (
    <div>
      {useTarget({
        qwik: () => {
          return <div>Hey</div>;
        },
        react: () => 'react',
        default: () => <span>Other Frameworks</span>,
      })}
    </div>
  );
}
```

- it should work anywhere: JSX, in the middle of a function, wrapping hooks, etc.
- it should be typed correctly: make sure the user provides a `default`


## 2- parse `useTarget`

In the JSX & Svelte parsers, whenever `useTarget` is detected, we need to add a new kind of block. This is going to be the hard part, and depends on where the `useTarget` is located.

I'll show one example: `useTarget` in the JSX render body:

```tsx
import { useState, useTarget, onUpdate } from '@builder.io/mitosis';

export default function MyComponent(props) {
  return (
    <div>
      {useTarget({
        qwik: () => {
          return <div>Hey</div>;
        },
        react: () => 'react',
        default: () => <span>Other Frameworks</span>,
      })}
    </div>
  );
}
```

I would expect a JSON to look like:

```json
{
  "name": "MyComponent",
  "@type": "@builder.io/mitosis/component",
  "children": [
    {
      "@type": "@builder.io/mitosis/node",
      "name": "div",
      "children": [
        {
          "@type": "@builder.io/mitosis/node",
          "name": "div",
          "children": [
            {
              "@type": "@builder.io/mitosis/conditional",
              "conditions": {
                "qwik": {
                  "@type": "@builder.io/mitosis/node",
                  "name": "div",
                  "properties": {
                    "_text": "Hey"
                  },
                  "children": []
                },
                "react": {
                  "@type": "@builder.io/mitosis/node",
                  "name": "div",
                  "properties": {
                    "_text": "\n      'react'\n    "
                  }
                },
                "default": {
                  "@type": "@builder.io/mitosis/node",
                  "name": "span",
                  "children": [
                    {
                      "@type": "@builder.io/mitosis/node",
                      "name": "div",
                      "properties": {
                        "_text": "Other Frameworks"
                      },
                    }
                  ]
                }
              }
            }
          ]
        }
      ]
    }
  ]
}
```

It will be a block with `"@type": "@builder.io/mitosis/conditional"`, and a `conditions` property that has all of the parsed conditions.

This should be:
- relatively straightforward to do for `useTarget` that return JSX (parse each individual condition and put it in the right place within the `children` JSX tree)
- hopefully also simple for `useTarget` wrapping hooks in the body of the component (this might require adding a `conditionals` array at the root of the `MitosisComponent` JSON schema
- more complex for `useTarget` that is located in an arbitrary code snippet, like:

```tsx
import { useStore, useTarget } from '@builder.io/mitosis';

export default function MyComponent(props) {
  const state = useStore({
    get val() {
      const conditionalVal = useTarget({
        qwik: () => {
          return 'qwik' + '10';
        },
        react: () => 'react',
        default: () => 'other',
      });

      return conditionalVal + '/abc';
    },
  });

  return <div>{state.val}</div>;
}
```

For this, probably an `inlined-conditionals` at the root of the component JSON as well that looks something like this:

```json
{
  "@type": "@builder.io/mitosis/component",
  "name": "MyComponent",
  "state": {
    "val": {
      "code": "get val() {\n  const conditionalVal = $$$CONDITIONAL_VALUE_0$$$;\n  return conditionalVal + '/abc';\n}",
      "type": "getter"
    }
  },
  "inlinedConditionals": {
    "0": {
      "qwik": {
        "code": "'qwik'+ 10",
        "type": "single"
      },
      "react": {
        "code": "'react'",
        "type": "single"
      },
      "default": {
        "code": "'other'",
        "type": "single"
      }
    }
  },
  "children": [
    {
      "@type": "@builder.io/mitosis/node",
      "name": "div",
      "children": [
        {
          "@type": "@builder.io/mitosis/node",
          "name": "div",
          "bindings": {
            "_text": {
              "code": "state.val",
              "type": "single"
            }
          }
        }
      ]
    }
  ] 
}
```

And the plugin from step 3 would inject the conditionals wherever there are `$$$CONDITIONAL_VALUE_X$$$` placeholders.

## 3- transform `conditional` blocks in generators

we would create a general-purpose plugin like this one: https://github.com/BuilderIO/mitosis/blob/c327e46d65fb6e01b5e91b3a9c20d6052b4c90ea/packages/core/src/helpers/plugins/process-code.ts#L61

it would iterate over the entire JSON object using `traverse()`, identify conditional blocks, and either inline or strip them away depending on which generator is calling the plugin.




## Notes

Since different uses of `useTarget` have varying degrees of implement complexity, it would be best to lay the foundation with the simplest one, and incrementally add more use-cases.

I think we should implement them in the following order:

- `useTarget` in the JSX render body
- `useTarget` in the component scope
- `useTarget` in JSX attributes/bindings/properties
- `useTarget` in `useState`/`useStore` expressions

**Comment by Lokprakash-babu** [comment]
@samijaber Please assign me to this enhancement. I would like to make a try on it.
